layout.js

```diff
@@ -1,144 +1,289 @@
+
```

Comment:
Is there any way we can test this to confirm it's working right?
Does it copy pixels always or just at end?